MINOR: improve test cleanup

mjsax opened the pull request:
- call close() on Metrics to join created threads

mjsax committed:
MINOR: imporve test cleanup
 - call close() on Metrics to join created threads

mjsax commented:
Call for review @enothereska @dguy 

Discovered by running test in infinite loop: fails eventually with
```
java.lang.OutOfMemoryError: unable to create new native thread

	at java.lang.Thread.start0(Native Method)
	at java.lang.Thread.start(Thread.java:714)
	at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:949)
	at java.util.concurrent.ThreadPoolExecutor.ensurePrestart(ThreadPoolExecutor.java:1590)
	at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:333)
	at java.util.concurrent.ScheduledThreadPoolExecutor.scheduleAtFixedRate(ScheduledThreadPoolExecutor.java:570)
	at org.apache.kafka.common.metrics.Metrics.<init>(Metrics.java:143)
	at org.apache.kafka.test.MockProcessorContext.<init>(MockProcessorContext.java:94)
	at org.apache.kafka.test.MockProcessorContext.<init>(MockProcessorContext.java:75)
	at org.apache.kafka.streams.state.internals.RocksDBSessionStoreSupplierTest.shouldNotBeLoggingEnabledStoreWhenLoggingNotEnabled(RocksDBSessionStoreSupplierTest.java:104)
	at sun.reflect.GeneratedMethodAccessor1.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:71)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

mjsax commented:
Retest this please.

dguy reviewed (commented):
General comment: Maybe put a `close` method on the `MockProcessorContext` and then it can close the metrics and probably any stores, too.
I don't like seeing this `context.baseMetrics().close()`

mjsax committed:
Damian's comment

mjsax commented:
@dguy I added `MockProcessorContext #close()` -- but it should not close any stores IMHO because it does not create any stores -- the `ProcessorStateManager` does close stores.

I also update `ProcessorStateManager#flush()` -- the parameter is never used (found this by chance) and thus removed it.

Also rebased to `trunk`. \cc @dguy @enothereska

enothereska commented on streams/src/test/java/org/apache/kafka/streams/state/internals/ChangeLoggingSegmentedBytesStoreTest.java:
```diff
@@ -37,6 +38,7 @@
 
 public class ChangeLoggingSegmentedBytesStoreTest {
 
+    private MockProcessorContext context;
```
Why isn't this `final` anymore?

mjsax committed:
Eno's comment

mjsax commented:
@eno Updated.

dguy commented:
>I also update ProcessorStateManager#flush() -- the parameter is never used (found this by chance) and thus removed it.

Nice one.

dguy commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java:
```diff
@@ -61,14 +62,15 @@ public void setUp() throws Exception {
 
     @Test
     public void shouldPutGetToFromCache() throws Exception {
+        context.close();
```
I don't understand why this is here? Shouldn't we have an `@After` where `context.close()` is called?

dguy commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBKeyValueStoreTest.java:
```diff
@@ -92,6 +92,7 @@ public void shouldUseCustomRocksDbConfigSetter() throws Exception {
         driver.setConfig(StreamsConfig.ROCKSDB_CONFIG_SETTER_CLASS_CONFIG, TheRocksDbConfigSetter.class);
         createKeyValueStore(driver.context(), Integer.class, String.class, false);
         assertTrue(TheRocksDbConfigSetter.called);
+        ((MockProcessorContext) driver.context()).close();
```
This won't get called if the test fails. I think it would be better to convert `driver` into a field and call `close` in an `@After`

dguy commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSessionStoreSupplierTest.java:
```diff
@@ -110,6 +111,7 @@ public void shouldNotBeLoggingEnabledStoreWhenLoggingNotEnabled() throws Excepti
         store.init(context, store);
         store.put(new Windowed<>("a", new SessionWindow(0, 10)), "b");
         assertTrue(logged.isEmpty());
+        context.close();
```
you should wrap in `try{...}finally{..}` so `context.close()` gets called when the test fails

mjsax committed:
Damian's comments

mjsax commented:
Updated.

dguy reviewed (approved):
Thanks @mjsax, LGTM

enothereska commented:
LGTM

mjsax commented:
Call for final review and merging @guozhangwang

guozhangwang commented:
retest this please

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/SinkNodeTest.java:
```diff
@@ -51,6 +51,8 @@ public void shouldThrowStreamsExceptionOnInputRecordWithInvalidTimestamp() {
             fail("Should have thrown StreamsException");
         } catch (final StreamsException ignored) {
+        } finally {
+            context.close();
```
Could we move `context` as this class's member variable, and then close the context in `@After`? All test cases seem to create the same object.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBKeyValueStoreTest.java:
```diff
@@ -86,19 +87,21 @@ public void setConfig(final String storeName, final Options options, final Map<S
-        final KeyValueStoreTestDriver<Integer, String> driver = KeyValueStoreTestDriver.create(Integer.class, String.class);
-        final MockProcessorContext context = (MockProcessorContext) driver.context();
-        final KeyValueStore<Integer, String> store = createStore(context, Integer.class, String.class, false, false);
+        final KeyValueStore<Integer, String> store = createStore(context, Integer.class, String.class, false);
```
I think we can just call `createKeyValueStore` and inline `createStore` inside `createKeyValueStore`. Also since all the calls in this class are the same, we can extract the `store` as a class variable.

mjsax replied:
We can make it a class member, but we cannot initialize globally because `shouldUseCustomRocksDbConfigSetter` uses a different `driver` setting to create the store.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSessionStoreSupplierTest.java:
```diff
@@ -106,10 +107,14 @@ public void shouldNotBeLoggingEnabledStoreWhenLoggingNotEnabled() throws Excepti
+            store.put(new Windowed<>("a", new SessionWindow(0, 10)), "b");
+            assertTrue(logged.isEmpty());
+        } finally {
+            context.close();
```
Since we call this already in `@After`, do we still need this finally block?

mjsax replied:
You are right. We can also refactor most vars as class members, too.

guozhangwang commented on streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java:
```diff
@@ -299,4 +296,8 @@ public ProcessorNode currentNode() {
         return currentNode;
     }
 
+    public void close() {
```
Add `@Override`.

mjsax replied:
This is no override.

guozhangwang replied:
Ah right, we only create the metrics object in this mock processor context, and in practice metrics is closed following another path. Thanks for the explanation.

guozhangwang reviewed (commented):
left some comments.

mjsax committed:
Github comments and some more cleanup

mjsax commented:
Updated this.

mjsax commented:
Retest this please.

guozhangwang reviewed (approved):
Merged to trunk.
